core/state-machine/src/ext.rs

```diff
@@ -152,15 +143,31 @@ where
 		self.storage_transaction = None;
 	}
 
+	/// Register the given extension for this instance.
+	pub fn register_extension<E: Any + Sized>(&mut self, ext: E) {
+		println!("REGISTER: {:?} {}", ext.type_id(), std::any::type_name::<E>());
+		self.extensions.register(ext);
+	}
+
+	/// Reset the internal state.
+	pub fn reset(&mut self) {
```

Comment:
That might be really flaky, can we add some documentation near the field declaration that you must reset the internal state here? To avoid situation in future, where people add/alter fields and forget that they need to be re-set.

Reply:
BTW why is that even better than creating externalities from scratch every time?

And also what's the logic behind WHAT should be cleared? Seems things like `overlay` are not. Maybe we can choose a better name for this? Seems it's only for randomness and transaction data?